Swift: detect hash functions with low # of iterations

karimhamdanali opened the pull request:
Using hash functions with less than 1,000 iterations is not secure. That scheme is vulnerable to password cracking attacks due to having an insufficient level of computational effort.

The rule currently supports all ciphers that the CryptoSwift API provides, but we can always extend it further if more cophers are added.

I'd appreciate a review of the query itself, the accompanying tests, and the associated documentation.

martin389 commented:
👋 Hi from docs, I've added this PR to the team board for a writer to review.

felicitymay commented:
Hi @karimhamdanali - thanks for asking for a docs team review of your pull request. 👋🏻 

Martin's followed our first responder notes and added this to our review board. Normally we expect a new query to have a technical review either complete or nearly complete before we start a docs review. This avoids duplicated effort because a technical review may change the focus of a query, so the user-facing content will change as a result of the review.

I'm going to remove the label and remove this pull request from our review board for now. Please do re-add the label when your confident that the technical review is nearly complete.

karimhamdanali commented:
Thanks @felicitymay for the clarification. I actually didn't know that! Will re-add the label once the PR is through the technical review.

geoffw0 commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.qhelp:
```diff
@@ -0,0 +1,17 @@
+  </overview>
+
+  <recommendation>
+    <p>Use sufficient number of iterations (i.e., greater than or equal 1000) for generating password-based keys.</p>
```
```suggestion
    <p>Use sufficient number of iterations (that is, greater than or equal 1000) for generating password-based keys.</p>
```
I believe "that is," is preferred over "i.e." (for readers with less experience in English).

Out of interest, I understand this comes from previous research but I'm interested to hear why the threshold we chose is 1,000, rather than say 100 or 10,000?  Is this recommendation reasonably future proof or is it something we might need to increase in future?

karimhamdanali replied:
Good question, the initial standard is 1000, but that number should increase with more CPU power (https://en.wikipedia.org/wiki/PBKDF2). The latest recommendation from OWASP is actually at least 120,000 iterations (https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html). Given that this number should increase over time, how do we best handle this?

geoffw0 replied:
If we want to increase it over time, I guess we should create some kind of issue to track this and preferably attach it to something (like an iteration or quarter) far in the future.

But it sounds like there's already a big gap between the current threshold of 1,000 and the OWASP recommendation of 120,000 iterations.  I guess we don't want to be too fussy - telling someone that 100,000 iterations isn't enough when the recommendation is 120,000 seems picky - but we do want to call out when code is risky e.g. because it follows advice from 20 years ago.  Is there any reason we shouldn't increase our threshold to somewhere in the 10,000 - 100,000 range?

geoffw0 replied:
... in fact reading the OWASP page, 120,000 appears to be the lowest of the recommendations they make.  So perhaps we _should_ set the threshold to 120,000???

karimhamdanali replied:
I'd say 120k should be the lowest threshold that we should generate warnings for. I'll update the queries and the associated documentation, examples, etc.

geoffw0 commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.qhelp:
```diff
@@ -0,0 +1,17 @@
+    <p>The following example shows a few cases of instantiating a password-based key. In the 'BAD' cases, the key is initialized with insufficient iterations, making it susceptible to password cracking attacks. In the 'GOOD' cases, the key is initialized with at least 1000 iterations, which protects the encrypted data against recovery.</p>
+    <sample src="InsufficientHashIterations.swift" />
+  </example>
+</qhelp>
```
I feel that this `.qhelp` badly needs a `<references>` section, with a link to something that explains why deliberately increasing the compute cost of these functions through iterations is considered a good use of resources (my understanding of the issue here).

karimhamdanali replied:
You're right! I forgot to add them to the initial PR. Added them now.

geoffw0 commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.ql:
```diff
@@ -0,0 +1,69 @@
+    exists(ClassOrStructDecl c, AbstractFunctionDecl f, CallExpr call, int arg |
+      c.getFullName() = ["PBKDF1", "PBKDF2"] and
+      c.getAMember() = f and
+      f.getName().matches("init(%iterations:%") and
```
This is slightly different from how we've usually matched function names before (listing all the variations explicitly).  I think I'm happy with it, it will be more robust if new versions of `init` with slightly different parameters are added in future.

karimhamdanali replied:
sgtm!

geoffw0 commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.ql:
```diff
@@ -0,0 +1,69 @@
+  DataFlow::PathNode sinkNode
+where config.hasFlowPath(sourceNode, sinkNode)
+select sinkNode.getNode(), sourceNode, sinkNode,
+  "The variable '" + sinkNode.getNode().toString() +
```
I think I would rephrase this as the number of iterations wasn't necessarily stored in a variable.  We could pull the number itself out from the source using `.getStringValue()` and describe it as "The value 123" (or whatever value it was).  Or we could simplify the message to something static like:
```
This value is an insufficient number of iterations for secure password hashing."
```

karimhamdanali replied:
I like your former suggestion, so I've changed the output message to something like 
`The value '999' is an insufficient number of iterations for secure password hashing.`

geoffw0 replied:
LGTM.

geoffw0 commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.ql:
```diff
@@ -0,0 +1,69 @@
+where config.hasFlowPath(sourceNode, sinkNode)
+select sinkNode.getNode(), sourceNode, sinkNode,
+  "The variable '" + sinkNode.getNode().toString() +
+    "' is an insufficient number of iterations, which is not secure for hash functions.",
```
```suggestion
    "' is an insufficient number of iterations, which is not secure for password hash functions.",
```
Many users will only read the query message despite our efforts with metadata and `.qhelp` and so forth - so I think its worth being as clear as possible here (while also being concise).

karimhamdanali replied:
I addressed that comment in my my latest commit to this PR.

geoffw0 replied:
Yep, thanks!

geoffw0 reviewed (commented):
Looks great, a few minor suggestions...

karimhamdanali commented:
Commit is now ready for docs review.

geoffw0 reviewed (commented):
I'm happy for this to be merged after the docs review. :+1:

martin389 commented:
Hi again from docs -- Adding this PR to the team board for a writer to review ✨

felicitymay commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.qhelp:
```diff
@@ -0,0 +1,22 @@
+  "qhelp.dtd">
+<qhelp>
+  <overview>
+    <p>Using hash functions with less than 1,000 iterations is not secure. That scheme is vulnerable to password cracking attacks due to having an insufficient level of computational effort.</p>
```
This query seems as if it's fairly closely related to the JavaScript query [Use of password hash with insufficient computational effort](https://codeql.github.com/codeql-query-help/javascript/js-insufficient-password-hash/).

I think that it might be worth reusing some of the explanation of why this is important.

karimhamdanali replied:
Agreed. I've updated the help text here with the relevant information from the JavaScript query. Thanks for the pointer to that query @felicitymay !

felicitymay commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.qhelp:
```diff
@@ -0,0 +1,22 @@
+  </overview>
+
+  <recommendation>
+    <p>Use sufficient number of iterations (that is, greater than or equal 120000) for generating password-based keys.</p>
```
We don't say how we arrived at this number. Is it taken from an OWASP recommendation, or somewhere else that we could refer to? I'm assuming that recommended minimums will gradually rise, so referring to a source where the current best practice is detailed will help this information stay up to date.

karimhamdanali replied:
Agreed. I've updated the text to state that we should follow the OWASP recommendation, with a note to the current recommendation of at least 120,000 iterations.

felicitymay commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.qhelp:
```diff
@@ -0,0 +1,22 @@
+  </recommendation>
+
+  <example>
+    <p>The following example shows a few cases of instantiating a password-based key. In the 'BAD' cases, the key is initialized with insufficient iterations, making it susceptible to password cracking attacks. In the 'GOOD' cases, the key is initialized with at least 120000 iterations, which protects the encrypted data against recovery.</p>
```
Minor text suggestion:

```suggestion
    <p>The following example shows a few cases where a password-based key is instantiated. In the 'BAD' cases, the key is initialized with insufficient iterations, making it susceptible to password cracking attacks. In the 'GOOD' cases, the key is initialized with at least 120,000 iterations, which protects the encrypted data against recovery.</p>
```

karimhamdanali replied:
fixed as suggested

felicitymay commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.swift:
```diff
@@ -0,0 +1,14 @@
+func encrypt() {
+	// ...
+
+	// BAD: Using insufficient (i.e., < 120,000) hash iterations keys for encryption
```
We write out "i.e." in full English for clarity:

```suggestion
	// BAD: Using insufficient (that is, < 120,000) hash iterations keys for encryption
```

karimhamdanali replied:
fixed as suggested

felicitymay commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.swift:
```diff
@@ -0,0 +1,14 @@
+	_ = try PKCS5.PBKDF1(password: getRandomArray(), salt: getRandomArray(), iterations: 90000, keyLength: 0)
+	_ = try PKCS5.PBKDF2(password: getRandomArray(), salt: getRandomArray(), iterations: 90000, keyLength: 0)
+
+	// GOOD: Using sufficient (i.e., >= 120,000) hash iterations keys for encryption
```
```suggestion
	// GOOD: Using sufficient (that is, >= 120,000) hash iterations keys for encryption
```

karimhamdanali replied:
fixed as suggested

felicitymay commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.swift:
```diff
@@ -0,0 +1,14 @@
+	_ = try PKCS5.PBKDF2(password: getRandomArray(), salt: getRandomArray(), iterations: 90000, keyLength: 0)
+
+	// GOOD: Using sufficient (i.e., >= 120,000) hash iterations keys for encryption
+	_ = try PKCS5.PBKDF1(password: getRandomArray(), salt: getRandomArray(), iterations: 120120, keyLength: 0)
+	_ = try PKCS5.PBKDF2(password: getRandomArray(), salt: getRandomArray(), iterations: 120120, keyLength: 0)
```
Judging by the OWASP reference, what is an adequate number of iterations seems to vary according to the algorithm you're using. It's not entirely clear to me looking at this example, whether or not this is adequate for the algorithm because I don't see which hashing algorithm is used.

This may be entirely unimportant - I'm a writer, not a developer 🙂

geoffw0 replied:
I think this is a good point.  We don't _implement_ the different recommended values for different algorithms, rather we just took the lowest recommendation and check the value used is at least that.  In practice this design is likely to catch the vast majority of problems and keeps the query simple and reliable - but there's no reason our _examples_ shouldn't set a good example by being fully OWASP compliant.

karimhamdanali replied:
Agreed. I've updated the test to be fully OWASP compliant :)

felicitymay commented on swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.ql:
```diff
@@ -0,0 +1,68 @@
+/**
+ * @name Insufficient hash iterations
+ * @description Using hash functions with less than 120,000 iterations is not secure, because that scheme leads to password cracking attacks due to having an insufficient level of computational effort.
```
Suggest:

```suggestion
 * @description Using hash functions with fewer than 120,000 iterations is insufficient to protect passwords because a cracking attack will require a low level of computational effort.
```

karimhamdanali replied:
fixed as suggested

felicitymay reviewed (commented):
Hi @karimhamdanali 👋🏻 

Many thanks for including query help, references and an example for this new query ✨ 

I've added a few suggestions and comments with the intention of making the information clearer for users. I'm hopeful that my suggestions don't change the intended meaning, but am aware that I have little technical context. Very happy to discuss any suggestions or comments.

karimhamdanali commented:
Thanks @felicitymay for the docs review! I think I've addressed all your comments so far. Please let me know if you have any further comments or suggestions before we can merge this PR. Thanks :)

felicitymay reviewed (approved):
Many thanks for the updates to the query help 💖

karimhamdanali commented:
Awesome! Thanks @felicitymay. @geoffw0, I think this one is ready to get merged then :) Thanks all for reviewing this PR.
